File: src/core/schema_tests/python/test_hand.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Unit tests for HandPoseT and related types in isaacteleop.schema.

HandPoseT is a FlatBuffers table that represents hand pose data:
- joints: HandJoints struct with a fixed-size poses array (length HandJoint.NUM_JOINTS; OpenXR order)

HandJoints is a struct with a fixed-size array of HandJointPose (length HandJoint.NUM_JOINTS).

HandJointPose is a struct containing:
- pose: The Pose (position and orientation)
- is_valid: Whether this joint data is valid
- radius: The radius of the joint (from OpenXR)

Timestamps are carried by HandPoseRecord, not HandPoseT.
"""

import pytest

from isaacteleop.schema import (
    DeviceDataTimestamp,
    HandJoint,
    HandJointPose,
    HandJoints,
    HandPoseRecord,
    HandPoseT,
    Point,
    Pose,
    Quaternion,
)


def test_hand_joint_enum_sentinels():
    """HandJoint ordinals match expected OpenXR-style layout."""
    assert HandJoint.PALM == 0
    assert HandJoint.WRIST == 1
    assert HandJoint.THUMB_TIP == 5
    assert HandJoint.LITTLE_TIP == 25
    assert HandJoint.NUM_JOINTS == 26


class TestHandJointPoseConstruction:
    """Tests for HandJointPose construction."""

    def test_default_construction(self):
        """Test default construction creates HandJointPose with default values."""
        joint_pose = HandJointPose()

        assert joint_pose is not None
        # Default pose values should be zero.
        assert joint_pose.pose.position.x == 0.0
        assert joint_pose.pose.position.y == 0.0
        assert joint_pose.pose.position.z == 0.0
        assert joint_pose.is_valid is False
        assert joint_pose.radius == 0.0

    def test_construction_with_values(self):
        """Test construction with position, orientation, is_valid, and radius."""
        position = Point(1.0, 2.0, 3.0)
        orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        pose = Pose(position, orientation)
        joint_pose = HandJointPose(pose, True, 0.01)

        assert joint_pose.pose.position.x == pytest.approx(1.0)
        assert joint_pose.pose.position.y == pytest.approx(2.0)
        assert joint_pose.pose.position.z == pytest.approx(3.0)
        assert joint_pose.is_valid is True
        assert joint_pose.radius == pytest.approx(0.01)


class TestHandJointPoseAccess:
    """Tests for HandJointPose property access."""

    def test_pose_access(self):
        """Test accessing pose property."""
        position = Point(1.5, 2.5, 3.5)
        orientation = Quaternion(0.1, 0.2, 0.3, 0.9)
        pose = Pose(position, orientation)
        joint_pose = HandJointPose(pose, True, 0.015)

        assert joint_pose.pose.position.x == pytest.approx(1.5)
        assert joint_pose.pose.orientation.w == pytest.approx(0.9)

    def test_is_valid_access(self):
        """Test accessing is_valid property."""
        pose = Pose(Point(), Quaternion())
        joint_pose = HandJointPose(pose, True, 0.0)

        assert joint_pose.is_valid is True

    def test_radius_access(self):
        """Test accessing radius property."""
        pose = Pose(Point(), Quaternion())
        joint_pose = HandJointPose(pose, False, 0.025)

        assert joint_pose.radius == pytest.approx(0.025)


class TestHandJointPoseRepr:
    """Tests for HandJointPose __repr__ method."""

    def test_repr(self):
        """Test __repr__ returns a meaningful string."""
        pose = Pose(Point(1.0, 2.0, 3.0), Quaternion(0.0, 0.0, 0.0, 1.0))
        joint_pose = HandJointPose(pose, True, 0.01)

        repr_str = repr(joint_pose)

        assert "HandJointPose" in repr_str
        assert "Pose" in repr_str


class TestHandJointsStruct:
    """Tests for HandJoints struct."""

    def test_poses_access(self):
        """Test accessing every joint slot via poses() method."""
        hand_joints = HandJoints()

        for i in range(HandJoint.NUM_JOINTS):
            joint = hand_joints.poses(i)
            assert joint is not None

    def test_poses_out_of_range(self):
        """Test that accessing out of range index raises IndexError."""
        hand_joints = HandJoints()

        with pytest.raises(IndexError):
            _ = hand_joints.poses(HandJoint.NUM_JOINTS)


class TestHandJointsRepr:
    """Tests for HandJoints __repr__ method."""

    def test_repr(self):
        """Test __repr__ returns a meaningful string."""
        hand_joints = HandJoints()

        repr_str = repr(hand_joints)
        assert "HandJoints" in repr_str


class TestHandPoseTConstruction:
    """Tests for HandPoseT construction and basic properties."""

    def test_default_construction(self):
        """Test default construction creates HandPoseT with pre-populated joints."""
        hand_pose = HandPoseT()

        assert hand_pose is not None
        assert hand_pose.joints is not None

    def test_parameterized_construction(self):
        """Test construction with joints."""
        joints = HandJoints()
        hand_pose = HandPoseT(joints)

        assert hand_pose.joints is not None


class TestHandPoseTRepr:
    """Tests for HandPoseT __repr__ method."""

    def test_repr_default(self):
        """Test __repr__ with default construction."""
        hand_pose = HandPoseT()

        repr_str = repr(hand_pose)
        assert "HandPoseT" in repr_str

    def test_repr_with_values(self):
        """Test __repr__ with joints set."""
        hand_pose = HandPoseT(HandJoints())

        repr_str = repr(hand_pose)
        assert "HandPoseT" in repr_str


class TestHandPoseRecordTimestamp:
    """Tests for HandPoseRecord with DeviceDataTimestamp."""

    def test_construction_with_timestamp(self):
        """Test HandPoseRecord carries DeviceDataTimestamp."""
        data = HandPoseT(HandJoints())
        ts = DeviceDataTimestamp(1000000000, 2000000000, 3000000000)
        record = HandPoseRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 1000000000
        assert record.timestamp.sample_time_local_common_clock == 2000000000
        assert record.timestamp.sample_time_raw_device_clock == 3000000000
        assert record.data is not None

    def test_default_construction(self):
        """Test default HandPoseRecord has no data."""
        record = HandPoseRecord()
        assert record.data is None
        assert record.timestamp is None

    def test_timestamp_fields(self):
        """Test all three DeviceDataTimestamp fields are accessible."""
        data = HandPoseT()
        ts = DeviceDataTimestamp(111, 222, 333)
        record = HandPoseRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 111
        assert record.timestamp.sample_time_local_common_clock == 222
        assert record.timestamp.sample_time_raw_device_clock == 333

File: src/core/schema_tests/cpp/test_hand.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Unit tests for the generated HandPose FlatBuffer message.

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>
#include <flatbuffers/flatbuffers.h>
#include <openxr/openxr.h>

// Include generated FlatBuffer headers.
#include <schema/hand_generated.h>
#include <schema/timestamp_generated.h>

#include <memory>
#include <type_traits>

// =============================================================================
// Compile-time verification of FlatBuffer field IDs.
// These ensure schema field IDs remain stable across changes.
// VT values are computed as: (field_id + 2) * 2.
// =============================================================================
#define VT(field) (field + 2) * 2
static_assert(core::HandPose::VT_JOINTS == VT(0));

// =============================================================================
// Compile-time verification of FlatBuffer field types.
// These ensure schema field types remain stable across changes.
// =============================================================================
#define TYPE(field) decltype(std::declval<core::HandPose>().field())
static_assert(std::is_same_v<TYPE(joints), const core::HandJoints*>);

// =============================================================================
// Compile-time verification of HandJointPose struct.
// =============================================================================
static_assert(std::is_trivially_copyable_v<core::HandJointPose>, "HandJointPose should be a trivially copyable struct");

// =============================================================================
// Compile-time verification of HandJoints struct.
// =============================================================================
static_assert(std::is_trivially_copyable_v<core::HandJoints>, "HandJoints should be a trivially copyable struct");

// HandJoints should contain exactly HandJoint::NUM_JOINTS HandJointPose entries.
static_assert(sizeof(core::HandJoints) == core::HandJoint_NUM_JOINTS * sizeof(core::HandJointPose),
              "HandJoints size must match HandJoint::NUM_JOINTS");

// =============================================================================
// OpenXR parity: core::HandJoint ordinals must match XrHandJointEXT (XR_EXT_hand_tracking).
// =============================================================================
static_assert(static_cast<int>(core::HandJoint_NUM_JOINTS) == XR_HAND_JOINT_COUNT_EXT,
              "HandJoint::NUM_JOINTS must match OpenXR XR_HAND_JOINT_COUNT_EXT");
static_assert(static_cast<int>(core::HandJoint_PALM) == static_cast<int>(XR_HAND_JOINT_PALM_EXT));
static_assert(static_cast<int>(core::HandJoint_WRIST) == static_cast<int>(XR_HAND_JOINT_WRIST_EXT));
static_assert(static_cast<int>(core::HandJoint_THUMB_METACARPAL) == static_cast<int>(XR_HAND_JOINT_THUMB_METACARPAL_EXT));
static_assert(static_cast<int>(core::HandJoint_THUMB_PROXIMAL) == static_cast<int>(XR_HAND_JOINT_THUMB_PROXIMAL_EXT));
static_assert(static_cast<int>(core::HandJoint_THUMB_DISTAL) == static_cast<int>(XR_HAND_JOINT_THUMB_DISTAL_EXT));
static_assert(static_cast<int>(core::HandJoint_THUMB_TIP) == static_cast<int>(XR_HAND_JOINT_THUMB_TIP_EXT));
static_assert(static_cast<int>(core::HandJoint_INDEX_METACARPAL) == static_cast<int>(XR_HAND_JOINT_INDEX_METACARPAL_EXT));
static_assert(static_cast<int>(core::HandJoint_INDEX_PROXIMAL) == static_cast<int>(XR_HAND_JOINT_INDEX_PROXIMAL_EXT));
static_assert(static_cast<int>(core::HandJoint_INDEX_INTERMEDIATE) ==
              static_cast<int>(XR_HAND_JOINT_INDEX_INTERMEDIATE_EXT));
static_assert(static_cast<int>(core::HandJoint_INDEX_DISTAL) == static_cast<int>(XR_HAND_JOINT_INDEX_DISTAL_EXT));
static_assert(static_cast<int>(core::HandJoint_INDEX_TIP) == static_cast<int>(XR_HAND_JOINT_INDEX_TIP_EXT));
static_assert(static_cast<int>(core::HandJoint_MIDDLE_METACARPAL) ==
              static_cast<int>(XR_HAND_JOINT_MIDDLE_METACARPAL_EXT));
static_assert(static_cast<int>(core::HandJoint_MIDDLE_PROXIMAL) == static_cast<int>(XR_HAND_JOINT_MIDDLE_PROXIMAL_EXT));
static_assert(static_cast<int>(core::HandJoint_MIDDLE_INTERMEDIATE) ==
              static_cast<int>(XR_HAND_JOINT_MIDDLE_INTERMEDIATE_EXT));
static_assert(static_cast<int>(core::HandJoint_MIDDLE_DISTAL) == static_cast<int>(XR_HAND_JOINT_MIDDLE_DISTAL_EXT));
static_assert(static_cast<int>(core::HandJoint_MIDDLE_TIP) == static_cast<int>(XR_HAND_JOINT_MIDDLE_TIP_EXT));
static_assert(static_cast<int>(core::HandJoint_RING_METACARPAL) == static_cast<int>(XR_HAND_JOINT_RING_METACARPAL_EXT));
static_assert(static_cast<int>(core::HandJoint_RING_PROXIMAL) == static_cast<int>(XR_HAND_JOINT_RING_PROXIMAL_EXT));
static_assert(static_cast<int>(core::HandJoint_RING_INTERMEDIATE) ==
              static_cast<int>(XR_HAND_JOINT_RING_INTERMEDIATE_EXT));
static_assert(static_cast<int>(core::HandJoint_RING_DISTAL) == static_cast<int>(XR_HAND_JOINT_RING_DISTAL_EXT));
static_assert(static_cast<int>(core::HandJoint_RING_TIP) == static_cast<int>(XR_HAND_JOINT_RING_TIP_EXT));
static_assert(static_cast<int>(core::HandJoint_LITTLE_METACARPAL) ==
              static_cast<int>(XR_HAND_JOINT_LITTLE_METACARPAL_EXT));
static_assert(static_cast<int>(core::HandJoint_LITTLE_PROXIMAL) == static_cast<int>(XR_HAND_JOINT_LITTLE_PROXIMAL_EXT));
static_assert(static_cast<int>(core::HandJoint_LITTLE_INTERMEDIATE) ==
              static_cast<int>(XR_HAND_JOINT_LITTLE_INTERMEDIATE_EXT));
static_assert(static_cast<int>(core::HandJoint_LITTLE_DISTAL) == static_cast<int>(XR_HAND_JOINT_LITTLE_DISTAL_EXT));
static_assert(static_cast<int>(core::HandJoint_LITTLE_TIP) == static_cast<int>(XR_HAND_JOINT_LITTLE_TIP_EXT));

// =============================================================================
// HandJointPose Tests
// =============================================================================
TEST_CASE("HandJointPose struct can be created and accessed", "[hand][struct]")
{
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::HandJointPose joint_pose(pose, true, 0.01f);

    SECTION("Pose values are accessible")
    {
        CHECK(joint_pose.pose().position().x() == 1.0f);
        CHECK(joint_pose.pose().position().y() == 2.0f);
        CHECK(joint_pose.pose().position().z() == 3.0f);
    }

    SECTION("is_valid is accessible")
    {
        CHECK(joint_pose.is_valid() == true);
    }

    SECTION("radius is accessible")
    {
        CHECK(joint_pose.radius() == Catch::Approx(0.01f));
    }
}

TEST_CASE("HandJointPose default construction", "[hand][struct]")
{
    core::HandJointPose joint_pose;

    // Default values should be zero/false.
    CHECK(joint_pose.pose().position().x() == 0.0f);
    CHECK(joint_pose.pose().position().y() == 0.0f);
    CHECK(joint_pose.pose().position().z() == 0.0f);
    CHECK(joint_pose.is_valid() == false);
    CHECK(joint_pose.radius() == 0.0f);
}

// =============================================================================
// HandJoints Tests
// =============================================================================
TEST_CASE("HandJoints struct has correct size", "[hand][struct]")
{
    // HandJoints should have exactly HandJoint::NUM_JOINTS entries.
    core::HandJoints joints;
    CHECK(joints.poses()->size() == static_cast<size_t>(core::HandJoint_NUM_JOINTS));
}

TEST_CASE("HandJoints can be accessed by index", "[hand][struct]")
{
    core::HandJoints joints;

    // Access first and last entries (returns pointers).
    const auto* first = (*joints.poses())[0];
    const auto* last = (*joints.poses())[static_cast<size_t>(core::HandJoint_NUM_JOINTS) - 1];

    // Default values should be zero.
    CHECK(first->pose().position().x() == 0.0f);
    CHECK(last->pose().position().x() == 0.0f);
}

// =============================================================================
// HandPoseT Tests
// =============================================================================
TEST_CASE("HandPoseT default construction", "[hand][native]")
{
    auto hand_pose = std::make_unique<core::HandPoseT>();

    // Default values.
    CHECK(hand_pose->joints == nullptr);
}

TEST_CASE("HandPoseT can store joints data", "[hand][native]")
{
    auto hand_pose = std::make_unique<core::HandPoseT>();

    // Create and set joints.
    hand_pose->joints = std::make_unique<core::HandJoints>();

    CHECK(hand_pose->joints->poses()->size() == static_cast<size_t>(core::HandJoint_NUM_JOINTS));
}

TEST_CASE("HandPoseT joints can be mutated via flatbuffers Array", "[hand][native]")
{
    auto hand_pose = std::make_unique<core::HandPoseT>();
    hand_pose->joints = std::make_unique<core::HandJoints>();

    // Create a joint pose.
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::HandJointPose joint_pose(pose, true, 0.015f);

    // Mutate first joint
    hand_pose->joints->mutable_poses()->Mutate(0, joint_pose);

    // Verify.
    const auto* first_joint = (*hand_pose->joints->poses())[0];
    CHECK(first_joint->pose().position().x() == 1.0f);
    CHECK(first_joint->pose().position().y() == 2.0f);
    CHECK(first_joint->pose().position().z() == 3.0f);
    CHECK(first_joint->is_valid() == true);
    CHECK(first_joint->radius() == Catch::Approx(0.015f));
}

TEST_CASE("HandPoseT serialization and deserialization", "[hand][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(4096);

    // Create HandPoseT with all fields set.
    auto hand_pose = std::make_unique<core::HandPoseT>();
    hand_pose->joints = std::make_unique<core::HandJoints>();

    // Set a few joint poses
    core::Point position(1.5f, 2.5f, 3.5f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose pose(position, orientation);
    core::HandJointPose joint_pose(pose, true, 0.02f);

    hand_pose->joints->mutable_poses()->Mutate(0, joint_pose);

    // Serialize.
    auto offset = core::HandPose::Pack(builder, hand_pose.get());
    builder.Finish(offset);

    // Deserialize.
    auto buffer = builder.GetBufferPointer();
    auto deserialized = flatbuffers::GetRoot<core::HandPose>(buffer);

    // Verify.
    CHECK(deserialized->joints()->poses()->size() == static_cast<size_t>(core::HandJoint_NUM_JOINTS));

    const auto* first_joint = (*deserialized->joints()->poses())[0];
    CHECK(first_joint->pose().position().x() == Catch::Approx(1.5f));
    CHECK(first_joint->pose().position().y() == Catch::Approx(2.5f));
    CHECK(first_joint->pose().position().z() == Catch::Approx(3.5f));
    CHECK(first_joint->is_valid() == true);
    CHECK(first_joint->radius() == Catch::Approx(0.02f));
}

TEST_CASE("HandPoseT can be unpacked from buffer", "[hand][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(4096);

    // Create and serialize.
    auto original = std::make_unique<core::HandPoseT>();
    original->joints = std::make_unique<core::HandJoints>();

    // Set multiple joint poses
    for (size_t i = 0; i < static_cast<size_t>(core::HandJoint_NUM_JOINTS); ++i)
    {
        core::Point position(static_cast<float>(i), static_cast<float>(i * 2), static_cast<float>(i * 3));
        core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
        core::Pose pose(position, orientation);
        core::HandJointPose joint_pose(pose, true, 0.01f + static_cast<float>(i) * 0.001f);
        original->joints->mutable_poses()->Mutate(i, joint_pose);
    }

    auto offset = core::HandPose::Pack(builder, original.get());
    builder.Finish(offset);

    // Unpack to HandPoseT.
    auto buffer = builder.GetBufferPointer();
    auto hand_pose_fb = flatbuffers::GetRoot<core::HandPose>(buffer);
    auto unpacked = std::make_unique<core::HandPoseT>();
    hand_pose_fb->UnPackTo(unpacked.get());

    // Check a few joints.
    const auto* joint_5 = (*unpacked->joints->poses())[5];
    CHECK(joint_5->pose().position().x() == Catch::Approx(5.0f));
    CHECK(joint_5->pose().position().y() == Catch::Approx(10.0f));
    CHECK(joint_5->pose().position().z() == Catch::Approx(15.0f));

    const size_t last_joint_index = static_cast<size_t>(core::HandJoint_NUM_JOINTS) - 1;
    const auto* joint_last = (*unpacked->joints->poses())[last_joint_index];
    CHECK(joint_last->pose().position().x() == Catch::Approx(static_cast<float>(last_joint_index)));
    CHECK(joint_last->pose().position().y() == Catch::Approx(static_cast<float>(last_joint_index * 2)));
    CHECK(joint_last->pose().position().z() == Catch::Approx(static_cast<float>(last_joint_index * 3)));
}

TEST_CASE("HandPoseT all joints can be set and verified", "[hand][native]")
{
    auto hand_pose = std::make_unique<core::HandPoseT>();
    hand_pose->joints = std::make_unique<core::HandJoints>();

    // Set every joint slot with a unique position.
    for (size_t i = 0; i < static_cast<size_t>(core::HandJoint_NUM_JOINTS); ++i)
    {
        core::Point position(static_cast<float>(i), 0.0f, 0.0f);
        core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
        core::Pose pose(position, orientation);
        core::HandJointPose joint_pose(pose, true, 0.01f);
        hand_pose->joints->mutable_poses()->Mutate(i, joint_pose);
    }

    // Verify all joints.
    for (size_t i = 0; i < static_cast<size_t>(core::HandJoint_NUM_JOINTS); ++i)
    {
        const auto* joint = (*hand_pose->joints->poses())[i];
        CHECK(joint->pose().position().x() == Catch::Approx(static_cast<float>(i)));
        CHECK(joint->is_valid() == true);
    }
}

// =============================================================================
// HandPoseRecord Tests (timestamp lives on the Record wrapper)
// =============================================================================
TEST_CASE("HandPoseRecord serialization with DeviceDataTimestamp", "[hand][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(4096);

    auto record = std::make_shared<core::HandPoseRecordT>();
    record->data = std::make_shared<core::HandPoseT>();
    record->data->joints = std::make_shared<core::HandJoints>();
    record->timestamp = std::make_shared<core::DeviceDataTimestamp>(1000000000LL, 2000000000LL, 3000000000LL);

    auto offset = core::HandPoseRecord::Pack(builder, record.get());
    builder.Finish(offset);

    auto deserialized = flatbuffers::GetRoot<core::HandPoseRecord>(builder.GetBufferPointer());

    CHECK(deserialized->timestamp()->available_time_local_common_clock() == 1000000000LL);
    CHECK(deserialized->timestamp()->sample_time_local_common_clock() == 2000000000LL);
    CHECK(deserialized->timestamp()->sample_time_raw_device_clock() == 3000000000LL);
    CHECK(deserialized->data()->joints()->poses()->size() == static_cast<size_t>(core::HandJoint_NUM_JOINTS));
}

TEST_CASE("HandPoseRecord can be unpacked with DeviceDataTimestamp", "[hand][flatbuffers]")
{
    flatbuffers::FlatBufferBuilder builder(4096);

    auto original = std::make_shared<core::HandPoseRecordT>();
    original->data = std::make_shared<core::HandPoseT>();
    original->data->joints = std::make_shared<core::HandJoints>();
    original->timestamp = std::make_shared<core::DeviceDataTimestamp>(111LL, 222LL, 333LL);

    auto offset = core::HandPoseRecord::Pack(builder, original.get());
    builder.Finish(offset);

    auto fb = flatbuffers::GetRoot<core::HandPoseRecord>(builder.GetBufferPointer());
    auto unpacked = std::make_shared<core::HandPoseRecordT>();
    fb->UnPackTo(unpacked.get());

    CHECK(unpacked->timestamp->available_time_local_common_clock() == 111LL);
    CHECK(unpacked->timestamp->sample_time_local_common_clock() == 222LL);
    CHECK(unpacked->timestamp->sample_time_raw_device_clock() == 333LL);
    CHECK(unpacked->data->joints->poses()->size() == static_cast<size_t>(core::HandJoint_NUM_JOINTS));
}
